core/async/src/test_loop/futures.rs

```diff
@@ -0,0 +1,146 @@
+//      and add DriveFutures as a handler. Finally, pass
+//      DelaySender<Arc<TestLoopTask>> as the &dyn FutureSpawner to any
+//      component that needs to spawn futures.
+//
+//      This causes any futures spawned during the test to end up as an event
+//      (an Arc<TestLoopTask>) in the test loop. The event will eventually be
+//      executed by the DriveFutures handler, which will drive the future
+//      until it is either suspended or completed. If suspended, then the waker
+//      of the future (called when the future is ready to resume) will place
+//      the Arc<TestLoopTask> event back into the test loop to be executed
+//      again.
+//
+//   2. A way to send a message to the TestLoop and expect a response as a
+//      future, which will resolve whenever the TestLoop handles the message.
+//      To support this, use MessageExpectingResponse<Request, Response> as the
+//      event type, and in the handler, call (event.responder)(result)
+//      (possibly asynchronously) to complete the future.
+//
+//      This is needed to support the AsyncSender interface, which is required
+//      by some components as they expect a response to each message. The way
+//      this is implemented is by implementing a conversion from
+//      DelaySender<MessageExpectingResponse<Request, Response>> to
+//      AsyncSender<Request, Response>.
+
+/// A message, plus a response callback. This should be used as the event type
+/// when testing an Actix component that's expected to return a result.
+///
+/// The response is used to complete the future that is returned by
+/// our `AsyncSender::send_async` implementation.
+pub struct MessageExpectingResponse<T, R> {
+    pub message: T,
+    pub responder: Box<dyn FnOnce(R) + Send>,
+}
+
+impl<T: Debug, R> Debug for MessageExpectingResponse<T, R> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("MessageWithResponder").field(&self.message).finish()
+    }
+}
+
+/// Function to create a pair of Future and MessageExpectingResponse, so that
+/// the future is resolved when the responder is called.
+fn create_future_for_message_response<T, R: Send + 'static>(
+    message: T,
+) -> (impl Future<Output = R>, MessageExpectingResponse<T, R>) {
+    let (sender, receiver) = oneshot::channel::<R>();
+    let future = async move { receiver.await.expect("Future was cancelled") };
+    let responder = Box::new(move |r| sender.send(r).ok().unwrap());
+    (future, MessageExpectingResponse { message, responder })
+}
+
+impl<
+        Message: 'static,
+        Response: Send + 'static,
+        Event: From<MessageExpectingResponse<Message, Response>> + 'static,
+    > messaging::CanSendAsync<Message, Response> for DelaySender<Event>
+{
+    fn send_async(&self, message: Message) -> BoxFuture<'static, Response> {
+        let (future, responder) = create_future_for_message_response(message);
```

Comment:
how about "response_sender"  & "response_receiver"?

Reply:
I inlined that function actually. In my previous iteration, I was implementing that future from scratch, and that helper function was more than a few lines. But then I realized I was reinventing a oneshot channel, so I deleted all that code and used oneshot channel instead, but then this helper function barely does anything.